src/Dropdown/Dropdown.js

```diff
@@ -2,57 +2,185 @@ import PropTypes from 'prop-types';
+
+    return options.filter(notSelected);
+  }
+
+  get selectedOption() {
+    const { options } = this.props;
+    const { selectedValue } = this.state;
+    const selected = option => sameValue(option.value, selectedValue);
+
+    return options.find(selected);
+  }
+
+  get selectedLabel() {
+    const { options } = this.props;
+    const { selectedValue } = this.state;
+    const selected = option => sameValue(option.value, selectedValue);
+    const selectedOption = options.find(selected);
 
     if (selectedOption) return selectedOption.label;
     if (options.length > 0) return options[0].label;
   }
 
+  focusActivator() {
+    this.activatorRef.current.focus();
+  }
+
+  closeOptions() {
+    this.setState(() => {
+      return { isOpen: false };
+    }, this.updateA11yMessage);
+  }
+
   handleClickOutside() {
-    this.setState({ isOpen: false });
+    this.closeOptions();
   }
 
   handleOptionClick = value => {
     const { onChange } = this.props;
-    this.setState({ isOpen: false, selectedValue: value }, () => {
-      onChange(value);
-    });
+
+    this.setState(
+      { isOpen: false, selectedValue: value, selectedIndex: INITIAL_INDEX },
+      () => {
+        onChange(value);
+        this.focusActivator();
+        this.updateA11yMessage();
+      }
+    );
   };
 
-  sameValue(value, otherValue) {
-    return value.toUpperCase() === otherValue.toUpperCase();
+  selectCurrentOption = () => {
+    const { onChange } = this.props;
+    const { selectedIndex } = this.state;
+    const options = this.optionsExceptSelected;
+    const selectedOption = options[selectedIndex];
+
+    if (selectedOption) {
+      this.setState(
+        () => ({
+          selectedValue: selectedOption.value,
+          selectedIndex: INITIAL_INDEX
+        }),
+        () => {
+          onChange(selectedOption.value);
+          this.activatorRef.current.focus();
+        }
+      );
+    }
+
+    this.closeOptions();
+  };
+
+  moveIndex(offset) {
+    const optionsCount = this.optionsExceptSelected.length;
+    const normalize = index => {
+      if (index < 0) {
+        return 0;
+      }
+
+      if (index >= optionsCount) {
+        return index - 1;
+      }
+
+      return index;
+    };
+
+    this.setState(prevState => ({
+      selectedIndex: normalize(prevState.selectedIndex + offset)
+    }));
   }
 
-  toggleOptions() {
+  handleKeyDown = evt => {
+    switch (evt.keyCode) {
+      case ARROW_UP:
+        evt.preventDefault();
+        this.moveIndex(DOWN);
+        break;
+      case ARROW_DOWN:
+        evt.preventDefault();
+        this.moveIndex(UP);
+        break;
+      case SPACE:
+        this.toggleOptions();
+        break;
+      case ENTER:
+        evt.preventDefault();
+        const { isOpen } = this.state;
+        isOpen ? this.selectCurrentOption() : this.toggleOptions();
+        break;
+      case ESC:
+        this.handleClickOutside();
```

Comment:
We can call directly to closeOptions() ? Not a stopper at all but maybe it's more readable at least to me. Actually, do we need handleClickOutside function if it just calls to closeOptions?

Reply:
Oww, of course, yes